### PR TITLE
Revive #1040 - spec/function.d

### PR DESCRIPTION
**Pull request by wilzbach**
Part of the revival of #1040.

### CODE REVIEW

File: spec/function.dd [hunk 1]
```diff
@@ -178,11 +178,65 @@ int foo();
...
+                //arr[num] = 1; // compile error
+                num = 2;        // has no side effect to the caller side
+                val.x = 3.14;   // ditto
+                return arr.length;
+            }
+            ---
+
+        $(P The maximum guarantee of `pure` is called "strong purity". It can enable
+            optimizations based on the fact
+            that a function is guaranteed to not mutate anyhing which isn't passed to it,
```
Comment (MetaLang) [inline spec/function.dd]: anyhing

File: spec/function.dd [hunk 2]
```diff
@@ -178,11 +178,65 @@ int foo();
...
+            and in cases where the compiler can guarantee that a pure function cannot
```
Comment (MetaLang) [inline spec/function.dd]: Very long run-on sentence. Consider breaking it up.

File: spec/function.dd [hunk 3]
```diff
@@ -178,11 +178,65 @@ int foo();
...
+            alter its arguments, it can enable full, functional purity (i.e. the guarantee
+            that the function will always return the same result for the same arguments).
+            To that end, a pure function:
```
Comment (MetaLang) [inline spec/function.dd]: This line is repeated again just below.

File: spec/function.dd [hunk 4]
```diff
@@ -2151,7 +2234,7 @@ void main()
 
 $(H2 $(LNAME2 interpretation, Compile Time Function Execution (CTFE)))
 
-    $(P Functions which are both portable and free of side-effects can be
+    $(P Functions which are both portable and free of side-effects to the global state can be
```
Comment (MetaLang) [inline spec/function.dd]: A better term might be "global side-effects".

File: spec/function.dd [hunk 5]
```diff
@@ -2329,6 +2412,84 @@ const int x = foo("1");
...
+$(H3 $(LNAME2 nogc-functions, No-GC Functions))
+
+        $(P No-GC functions are functions marked with the $(D @nogc) attribute.
+            Those functions do not allocate memory on the GC heap,
+            through the following language features:
+        )
+
+        $(UL
+        $(LI $(DDSUBLINK expression, ArrayLiteral, constructing an array) on the heap)
+        $(LI resizing an array with writing its $(D .length) property)
```
Comment (MetaLang) [inline spec/function.dd]: "with writing its" -> "by writing to its"

File: spec/function.dd [hunk 6]
```diff
@@ -2362,6 +2523,10 @@ $(H3 $(LNAME2 safe-functions, Safe Functions))
         $(LI Cannot access $(D __gshared) variables.)
         )
 
+        $(P With the array indexing and slicing, an out of bounds access
```
Comment (MetaLang) [inline spec/function.dd]: This might be better written as "when indexing or slicing an array".

**Comment by wilzbach** [comment]
Addressed the review and found the stray parens.

File: spec/function.dd [hunk 7]
```diff
@@ -2238,7 +2317,7 @@ int countTen(int x)
     return x;
 }
 
-static assert(countTen(6) == 6); // OK
+static assert(countTen(6) == 6);    // OK
```
Comment (MetaLang) [inline spec/function.dd]: space

**Pull request merged**